oio/xcute/orchestrator.py

```diff
@@ -0,0 +1,606 @@
+                        job_id, exc)
+
+    def handle_running_job(self, job_info):
+        """
+            Read the job's configuration
+            and get its tasks before dispatching it
+        """
+        if job_info['tasks']['all_sent']:
+            return
+
+        job_id = job_info['job']['id']
+        job_type = job_info['job']['type']
+        last_task_id = job_info['tasks']['last_sent']
+        total_marker = job_info['tasks']['total_marker']
+        job_config = job_info['config']
+        job_params = job_config['params']
+        job_class = JOB_TYPES[job_type]
+        job = job_class(self.conf, logger=self.logger)
+        job_tasks = job.get_tasks(job_params, marker=last_task_id)
+
+        tasks_counter = None
+        if job_info['tasks']['is_total_temp']:
+            tasks_counter = job.get_total_tasks(
+                job_params, marker=total_marker)
+
+        thread_args = (job_id, job_type, job_config, job_tasks)
+        dispatch_thread = threading.Thread(
+            target=self.dispatch_job,
+            args=thread_args)
+        dispatch_thread.start()
+
+        self.threads[dispatch_thread.ident] = dispatch_thread
+
+        if tasks_counter is None:
+            return
+
+        total_tasks_thread = threading.Thread(
+            target=self.get_job_total_tasks,
+            args=(job_id, tasks_counter))
+        total_tasks_thread.start()
+
+    def dispatch_job(self, job_id, job_type, job_config, job_tasks):
+        """
+            Dispatch all of a job's tasks
+        """
+
+        self.logger.info('Start dispatching job (job_id=%s)', job_id)
+
+        try:
+            beanstalkd_workers = self.get_beanstalkd_workers()
+
+            tasks_per_second = job_config['tasks_per_second']
+            tasks_batch_size = job_config['tasks_batch_size']
+
+            tasks_run_time = 0
+            batch_per_second = tasks_per_second / float(
+                tasks_batch_size)
+            # The backend must have the tasks in order
+            # to know the last task sent
+            tasks = OrderedDict()
+            for task_id, task_payload in job_tasks:
+                if not self.running:
+                    break
+
+                tasks[task_id] = task_payload
+                if len(tasks) < tasks_batch_size:
+                    continue
+
+                tasks_run_time = ratelimit(
+                    tasks_run_time, batch_per_second)
+
+                sent = self.dispatch_tasks(
+                    beanstalkd_workers,
+                    job_id, job_type, job_config, tasks)
+                if sent:
+                    job_status, exc = self.handle_backend_errors(
+                        self.backend.update_tasks_sent, job_id, tasks.keys())
+                    tasks.clear()
+                    if exc is not None:
+                        self.logger.warn(
+                            '[job_id=%s] Sent tasks has not been updated: %s',
```

Comment:
Sent tasks counter?

Reply:
It's not just the counter that hasn't been updated. We also keep all the IDs of the sent tasks (until they are processed).